File: app/src/main/java/ca/judacribz/gainzassist/models/db/WorkoutRepo.java
package ca.judacribz.gainzassist.models.db;

import static ca.judacribz.gainzassist.util.firebase.Database.addWorkoutSessionFirebase;
import static ca.judacribz.gainzassist.util.firebase.Database.deleteWorkoutFirebase;

import android.app.Application;
import android.content.Context;

import androidx.lifecycle.LiveData;

import com.google.firebase.database.DataSnapshot;

import java.util.List;

import ca.judacribz.gainzassist.interfaces.OnWorkoutReceivedListener;
import ca.judacribz.gainzassist.models.Exercise;
import ca.judacribz.gainzassist.models.ExerciseSet;
import ca.judacribz.gainzassist.models.Session;
import ca.judacribz.gainzassist.models.Workout;

public class WorkoutRepo {
    static private WorkoutDao workoutDao;
    static private ExerciseDao exerciseDao;
    static private SetDao setDao;
    static private SessionDao sessionDao;
    private DataSnapshot workoutShot;



    public enum TableTxn {
        WORKOUTS_TXN,
        EXERCISES_TXN,
        SESSIONS_TXN,
        SETS_TXN
    }

    public enum RepoTask {
        GET_WORKOUT,
        GET_EXERCISES,

        INSERT_EXERCISE,
        INSERT_WORKOUT,
        INSERT_SESSION,
        INSERT_SET,

        UPDATE_WORKOUT,
        UPDATE_EXERCISE,
        UPDATE_EXERCISE_WEIGHT,
        UPDATE_SET,

        DELETE_ALL_WORKOUTS,
        DELETE_WORKOUT,
        DELETE_EXERCISE,
        DELETE_SET
    }

    public WorkoutRepo(Application app) {
        WorkoutDatabase db = WorkoutDatabase.getDatabase(app);

        workoutDao = db.workoutDao();
        exerciseDao = db.exerciseDao();
        setDao = db.setDao();
        sessionDao = db.sessionDao();
    }


    // CRUD functions
    // ============================================================================================
    // CREATE
    // --------------------------------------------------------------------------------------------
    public void insertWorkout(Workout workout) {
        new Thread(() -> {
            workoutDao.insert(workout);
            insertExercise(workout.exercises.toArray(new Exercise[0]));
        }).start();
    }

    void insertExercise(Exercise...exercises) {
        for (Exercise ex : exercises) {
            exerciseDao.insert(ex);
        }
    }

    public void insertSession(Session session, boolean toFireBase) {
        new Thread(() -> {
            sessionDao.insert(session);
            for (Exercise ex : session.getSessionExs()) {
                insertSet(ex.getFinishedSetsList().toArray(new ExerciseSet[0]));
            }
        }).start();

        if (toFireBase) {
            addWorkoutSessionFirebase(session);
        }
    }

    static void insertSet(ExerciseSet...exerciseSets) {
        for (ExerciseSet set : exerciseSets) {
            new Thread(() -> setDao.insert(set)).start();
        }
    }
    // --------------------------------------------------------------------------------------------


    // RETRIEVE
    // --------------------------------------------------------------------------------------------
    LiveData<List<Workout>> getAllWorkoutsLive() {
        return workoutDao.getAll();
    }
    LiveData<List<Session>> getAllSessionsLive() {
        return sessionDao.getAll();
    }

    LiveData<List<ExerciseSet>> getAllSetsLive() {
        return setDao.getAll();
    }

    LiveData<Workout> getWorkout(long id) {
        return workoutDao.get(id);
    }
    OnWorkoutReceivedListener onWorkoutReceivedListener;
    private void setOnWorkoutReceivedListener(OnWorkoutReceivedListener onWorkoutReceivedListener) {
        this.onWorkoutReceivedListener = onWorkoutReceivedListener;
    }
    void getWorkoutFromName(Context context, String name) {
        new Thread(() -> {
            setOnWorkoutReceivedListener((OnWorkoutReceivedListener) context);
            Workout workout = workoutDao.getFromName(name);

            for (Exercise exercise : exerciseDao.getFromWorkout(workout.getId())) {
                exercise.setSetsList(null);
                workout.addExercise(exercise);
            }

            onWorkoutReceivedListener.onWorkoutsReceived(workout);
            setOnWorkoutReceivedListener(null);
        }).start();

    }

    LiveData<List<Exercise>> getExercisesFromWorkout(long workoutId) {
        return exerciseDao.getLiveFromWorkout(workoutId);
    }

    LiveData<Exercise> getExercise(long id) {
        return exerciseDao.getLive(id);
    }

    LiveData<List<String>> getAllUniqueExerciseNames() {
        return exerciseDao.getAllUniqueNames();
    }

    LiveData<List<ExerciseSet>> getSetsFromExercise(long exerciseId) {
        return setDao.getLiveFromExercise(exerciseId);
    }
    // --------------------------------------------------------------------------------------------


    // UPDATE
    // --------------------------------------------------------------------------------------------
    public void updateWorkout(Workout workout) {
        new Thread(() -> workoutDao.update(workout)).start();
        updateExercise(workout.exercises.toArray(new Exercise[0]));
    }

    void updateExercise(Exercise...exercises) {
        for (Exercise ex : exercises) {
            new Thread(() -> {
                if (exerciseDao.get(ex.getId()) == null) {
                    insertExercise(ex);
                } else {
                    exerciseDao.update(ex);
                }
            }).start();
        }
    }
    // --------------------------------------------------------------------------------------------


    // DELETE
    // --------------------------------------------------------------------------------------------
    void deleteAllWorkouts() {
        new Thread(() -> workoutDao.deleteAll()).start();
    }

    public void deleteWorkout(Workout workout) {
        new Thread(() -> workoutDao.delete(workout)).start();
        deleteWorkoutFirebase(workout.name);
    }

    void deleteWorkout(String workoutName) {
        new Thread(() -> workoutDao.delete(workoutName)).start();
        deleteWorkoutFirebase(workoutName);
    }

    void deleteExercise(Exercise...exercise) {
        new Thread(() -> exerciseDao.delete(exercise)).start();
    }

    void deleteSet(ExerciseSet exerciseSet) {
        new Thread(() -> setDao.delete(exerciseSet)).start();
    }
    // --------------------------------------------------------------------------------------------
    // ============================================================================================
}

File: app/src/main/java/ca/gainzassist/models/db/WorkoutRepo.kt
package ca.gainzassist.models.db

import android.app.Application
import android.content.Context
import androidx.lifecycle.LiveData
import ca.gainzassist.domain.usecase.workout.CalculateNextExerciseWeightUseCase
import ca.gainzassist.interfaces.OnWorkoutReceivedListener
import ca.gainzassist.models.Exercise
import ca.gainzassist.models.ExerciseSet
import ca.gainzassist.models.Session
import ca.gainzassist.models.Workout
import ca.gainzassist.util.firebase.Database.addWorkoutSessionFirebase
import ca.gainzassist.util.firebase.Database.deleteWorkoutFirebase
import com.google.firebase.database.DataSnapshot

class WorkoutRepo(app: Application) {
    private var workoutShot: DataSnapshot? = null
    var onWorkoutReceivedListener: OnWorkoutReceivedListener? = null

    enum class TableTxn {
        WORKOUTS_TXN,
        EXERCISES_TXN,
        SESSIONS_TXN,
        SETS_TXN
    }

    enum class RepoTask {
        GET_WORKOUT,
        GET_EXERCISES,
        INSERT_EXERCISE,
        INSERT_WORKOUT,
        INSERT_SESSION,
        INSERT_SET,
        UPDATE_WORKOUT,
        UPDATE_EXERCISE,
        UPDATE_EXERCISE_WEIGHT,
        UPDATE_SET,
        DELETE_ALL_WORKOUTS,
        DELETE_WORKOUT,
        DELETE_EXERCISE,
        DELETE_SET
    }

    init {
        val db = WorkoutDatabase.getDatabase(app)
        workoutDao = db.workoutDao()
        exerciseDao = db.exerciseDao()
        setDao = db.setDao()
        sessionDao = db.sessionDao()
    }

    // CRUD functions
    // ============================================================================================
    // CREATE
    // --------------------------------------------------------------------------------------------
    fun insertWorkout(workout: Workout) {
        Thread {
            workoutDao.insert(workout)
            insertExercise(*workout.exercises.toTypedArray())
        }.start()
    }

    fun insertExercise(vararg exercises: Exercise) {
        for (ex in exercises) {
            exerciseDao.insert(ex)
        }
    }

    fun insertSession(session: Session, toFireBase: Boolean) {
        Thread {
            sessionDao.insert(session)
            val calculateNextWeight = CalculateNextExerciseWeightUseCase()
            for (ex in session.sessionExs) {
                insertSet(*ex.getFinishedSetsList().toTypedArray())
                if (ex.setsType == Exercise.SetsType.MAIN_SET) {
                    val nextWeight = calculateNextWeight.invoke(ex, ex.getFinishedSetsList())
                    exerciseDao.updateWeight(nextWeight, ex.id)
                }
            }
        }.start()

        if (toFireBase) {
            addWorkoutSessionFirebase(session)
        }
    }

    // --------------------------------------------------------------------------------------------

    // RETRIEVE
    // --------------------------------------------------------------------------------------------
    fun getAllWorkoutsLive(): LiveData<List<Workout>> {
        return workoutDao.getAll()
    }

    fun getAllSessionsLive(): LiveData<List<Session>> {
        return sessionDao.getAll()
    }

    fun getAllSetsLive(): LiveData<List<ExerciseSet>> {
        return setDao.getAll()
    }

    fun getWorkout(id: Long): LiveData<Workout> {
        return workoutDao.get(id)
    }

    fun getWorkoutFromName(context: Context, name: String) {
        Thread {
            onWorkoutReceivedListener = context as OnWorkoutReceivedListener
            val workout = workoutDao.getFromName(name) ?: return@Thread

            for (exercise in exerciseDao.getFromWorkout(workout.id)) {
                workout.addExercise(exercise)
            }

            onWorkoutReceivedListener?.onWorkoutsReceived(workout)
            onWorkoutReceivedListener = null
        }.start()
    }

    fun getExercisesFromWorkout(workoutId: Long): LiveData<List<Exercise>> {
        return exerciseDao.getLiveFromWorkout(workoutId)
    }

    fun getExercise(id: Long): LiveData<Exercise> {
        return exerciseDao.getLive(id)
    }

    fun getAllUniqueExerciseNames(): LiveData<List<String>> {
        return exerciseDao.getAllUniqueNames()
    }

    fun getSetsFromExercise(exerciseId: Long): LiveData<List<ExerciseSet>> {
        return setDao.getLiveFromExercise(exerciseId)
    }
    // --------------------------------------------------------------------------------------------

    // UPDATE
    // --------------------------------------------------------------------------------------------
    fun updateWorkout(workout: Workout) {
        Thread { workoutDao.update(workout) }.start()
        updateExercise(*workout.exercises.toTypedArray())
    }

    fun updateExercise(vararg exercises: Exercise) {
        for (ex in exercises) {
            Thread {
                if (exerciseDao.get(ex.id) == null) {
                    insertExercise(ex)
                } else {
                    exerciseDao.update(ex)
                }
            }.start()
        }
    }
    // --------------------------------------------------------------------------------------------

    // DELETE
    // --------------------------------------------------------------------------------------------
    fun deleteAllWorkouts() {
        Thread { workoutDao.deleteAll() }.start()
    }

    fun deleteWorkout(workout: Workout) {
        Thread { workoutDao.delete(workout) }.start()
        deleteWorkoutFirebase(workout.name)
    }

    fun deleteWorkout(workoutName: String) {
        Thread { workoutDao.delete(workoutName) }.start()
        deleteWorkoutFirebase(workoutName)
    }

    fun deleteExercise(vararg exercise: Exercise) {
        Thread { exerciseDao.delete(*exercise) }.start()
    }

    fun deleteSet(exerciseSet: ExerciseSet) {
        Thread { setDao.delete(exerciseSet) }.start()
    }
    // --------------------------------------------------------------------------------------------
    // ============================================================================================

    companion object {
        private lateinit var workoutDao: WorkoutDao
        private lateinit var exerciseDao: ExerciseDao
        private lateinit var setDao: SetDao
        private lateinit var sessionDao: SessionDao

        @JvmStatic
        fun insertSet(vararg exerciseSets: ExerciseSet) {
            for (set in exerciseSets) {
                Thread { setDao.insert(set) }.start()
            }
        }
    }
}
